From c03dbf36f0b085b37736fefcf37cc6b5766ce73e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Old=C5=99ich=20Jedli=C4=8Dka?= <oldium.pro@gmail.com>
Date: Sun, 28 Jan 2024 11:47:51 +0100
Subject: [PATCH] Fix killing of child process of clevisloop
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The usage of `ps -l` is fragile and does not always return the expected
set of columns. It differs even between several BusyBox versions and
operating systems. Use `ps -o pid,ppid` to get the expected set of columns
instead.

Also kill parent first before its children to prevent spawning another
command from clevisloop itself.

Fixes #448

Signed-off-by: Oldřich Jedlička <oldium.pro@gmail.com>
---
 src/initramfs-tools/scripts/local-bottom/clevis.in | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/initramfs-tools/scripts/local-bottom/clevis.in b/src/initramfs-tools/scripts/local-bottom/clevis.in
index fb89e35c..fff3d758 100755
--- a/src/initramfs-tools/scripts/local-bottom/clevis.in
+++ b/src/initramfs-tools/scripts/local-bottom/clevis.in
@@ -34,8 +34,10 @@ esac
 [ -s /run/clevis.pid ] || exit 0
 
 pid=$(cat /run/clevis.pid)
-ps -l | awk -v pid="$pid" '$4==pid { system("kill " $3) }'
-kill "$pid"
+child_pids=$(ps -o pid,ppid | awk -v pid="$pid" '$2==pid { print $1 }')
+for kill_pid in $pid $child_pids; do
+	kill "$kill_pid"
+done
 
 # Not really worried about downing extra interfaces: they will come up
 # during the actual boot. Might make this configurable later if needed.
